Adding the ability to use machine translation providers to the copy feature

IsaiahPaget opened the pull request:
#102

Created a trait that handles translation logic
Updated the copy flow to use a popup to confirm the copy and select provider.

The feature is working, but I would just appreciate some feedback on the UI and maybe the new APIs I made from the perspective of a developer using this plugin.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added auto-translate support when copying content between locales, with a selectable translation provider (Google, DeepL) and a “none” option.
  * Integrated provider-based auto-translation into text, textarea, markdown, rich editor, repeaters, and blocks.
  * Added configuration for providers, default provider, and a whitelist of auto-translatable fields.

* **UI**
  * Added translation method/provider selector modal and improved provider label presentation.
  * Refined multilingual styling, including language code display and loading indicator layering; improved dropdown layout consistency.

* **Tests**
  * Added unit tests covering provider resolution, translation flows, and whitelist utilities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Add auto-translation support with Google & DeepL providers

Commit:
Addressing feedback from AIC

better errors using, select instead of radio group

fix the richeditor

Fixed mediafinder issue, updated config, updated the way whitelist is
defined

AIC-BV commented:
https://github.com/user-attachments/assets/3e805a0f-300d-4980-b374-ebf81b70bad0

When swapping languages, (not saving in between), the translation disappears and the text is as if the translation never happened but a simple copy instead

AIC-BV commented:
Since Deepl is not set up on my end, it should be in the list
Do I have to override the config file at `root/config/wintercms/wn-translate-plugin/config.php` ?

AIC-BV commented:
I put
```
'autoTranslateWhiteList' => [
        'slug',
        // 'name',
        // 'content',
    ],
```
If I understand it correctly, only field 'slug' can be translated?
But I can still translate every field.

Also wondering wether a blacklist wouldn't make more sense than a whitelist

Commit:
update config

IsaiahPaget commented:
@AIC-BV I have addressed the issues that you have brought up here, and updated the config file to have better documentation of the feature, and added errors that are more clear, hope this work, let me know if you have any more issues!

LukeTowers commented:
> The only problem I have right now is that I modified translate/config/config.php but I have to override it somewhere in my own files to avoid composer update overwriting my changes

@AIC-BV see https://wintercms.com/docs/v1.2/docs/plugin/settings#file-based-configuration, you would just put your overrides in `myproject/config/winter/translate/config.php`

If you're happy with this PR's functionality as it stands right now @AIC-BV let me know and I'll do a final cleanup pass on the code & UI before I approve it for merging and we send you the invoice.

Commit:
cleaning up the concept of 'standard' translation and renaming it to 'none'

AIC-BV commented:
@LukeTowers looks ready to me
The popup UI could perhaps be enhanced but the functionality itself seems to be perfectly working

Thanks @IsaiahPaget for the work

AIC-BV commented:
How do I get the latest version?
I did `"winter/wn-translate-plugin": "dev-main#ed7ab05d40a79acd6ff76a7108050a7fc3d1d613"`, in composer but I don't think this was the latest?

AIC-BV commented:
Saves so much time 🤩

AIC-BV commented:
There's a bug using this PR at backendforms type fileupload
<img width="1013" height="1025" alt="image" src="https://github.com/user-attachments/assets/816bc83a-218c-4c7b-898a-8a8084349847" />

You can no longer fill in the title/description caption
```
photos:
        label: Opmeter
        type: fileupload
        mode: image
        useCaption: true
        attachOnUpload: true
        thumbOptions:
            mode: crop
            offset:
                - 0
                - 0
            quality: 90
            sharpen: 0
            interlace: false
            extension: auto
        span: auto
   ```

Commit:
Using a proper popup

I wasn't using proper popups before and this resulted in the copy button
submitting forms that it was inside, which would close the file upload
popup and relation popup, and just generally cause issues.

AIC-BV commented on providers/GoogleTranslateProvider.php:
```diff
@@ -0,0 +1,45 @@
+
+        $endpoint = rtrim($this->config['url'], '/') . '?' . $query;
+
+        $response = Http::get($endpoint);
```
```suggestion
        $endpoint = rtrim($this->config['url'], '/');
        
        $response = Http::withBody($query, 'application/x-www-form-urlencoded')->post($endpoint);
```

AIC-BV reviewed (changes requested):
GoogleTranslateProvider should send translation parameters in the POST body instead of the GET query string.

This fixes auto-translation failing with a Google 400 Bad Request (generic HTML "malformed or illegal request" page) on longer fields. Rich-editor/textarea content, once url-encoded into the GET query string, exceeded Google's ~8KB request-line limit and was rejected by Google's frontend before reaching the Translation API. Moving the payload to the POST body removes the URL-length ceiling; request format and response parsing are unchanged.

IsaiahPaget committed:
Update providers/GoogleTranslateProvider.php

Co-authored-by: Meindert <89913092+AIC-BV@users.noreply.github.com>

AIC-BV commented:
We run this branch in production with language-country locale codes (`nl-BE`, `fr-BE`) and hit two small UI issues: the locale button's fixed 44px width can't fit longer codes, and the full-height `.ml-dropdown` wrappers block clicks on the UI underneath (e.g. the mediafinder upload button) while a dropdown is open.

Fixes for both are in IsaiahPaget/wn-translate-plugin#1, targeting this PR's branch so they can land here.

Before:
<img width="212" height="91" alt="image" src="https://github.com/user-attachments/assets/862c8b9a-4af1-4a5e-ae1c-271a99a56366" />
After:
<img width="201" height="112" alt="image" src="https://github.com/user-attachments/assets/777a8545-6815-4259-9c88-361bfb0e4fe7" />

LukeTowers commented:
Closing in favour of #115
